<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.14: http://docutils.sourceforge.net/" />
<title>PGAPy: Python Wrapper for PGAPack Parallel Genetic Algorithm Library</title>
<meta name="author" content="Ralf Schlatterbeck &lt;rsc&#64;runtux.com&gt;" />
<style type="text/css">

/*
:Author: David Goodger
:Contact: goodger@users.sourceforge.net
:Date: $Date: 2007-12-21 16:51:22 +0100 (Fri, 21 Dec 2007) $
:Version: $Revision: 4316 $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.
*/

/* "! important" is used here to override other ``margin-top`` and
   ``margin-bottom`` styles that are later in the stylesheet or 
   more specific.  See <http://www.w3.org/TR/CSS1#the-cascade>. */
.first {
  margin-top: 0 ! important }

.last {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

a.reference img {
  border-width: 0px ;
  }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em }

div.footer, div.header {
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin-left: 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font-family: serif ;
  font-size: 100% }

pre.line-block {
  font-family: serif ;
  font-size: 100% }

pre.literal-block, pre.doctest-block {
  margin-left: 2em ;
  margin-right: 2em ;
  background-color: #eeeeee }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

table.citation {
  border-left: solid thin gray }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid thin black }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

th.docinfo-name, th.field-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

tt.docutils {
  background-color: #eeeeee }

ul.auto-toc {
  list-style-type: none }

body {
  background-color: #F0F0FF }

p img {
  vertical-align: text-top }


</style>
</head>
<body>
<div class="document" id="pgapy-python-wrapper-for-pgapack-parallel-genetic-algorithm-library">
<h1 class="title">PGAPy: Python Wrapper for PGAPack Parallel Genetic Algorithm Library</h1>
<table class="docinfo" frame="void" rules="none">
<col class="docinfo-name" />
<col class="docinfo-content" />
<tbody valign="top">
<tr><th class="docinfo-name">Author:</th>
<td>Ralf Schlatterbeck &lt;<a class="reference external" href="mailto:rsc&#64;runtux.com">rsc&#64;runtux.com</a>&gt;</td></tr>
</tbody>
</table>
<p>News: This version wraps the Differential Evolution method (that's quite
an old method but is newly implemented in pgapack).</p>
<p>PGAPy is a wrapper for PGAPack, the parallel genetic algorithm library
(see <a class="reference external" href="https://github.com/schlatterbeck/pgapack/blob/master/README.rst">PGAPack Readme</a>), a powerfull genetic algorithm library by
D. Levine, Mathematics and Computer Science Division Argonne National
Laboratory. The library is written in C. PGAPy wraps this library for
use with Python. The original PGAPack library is already quite old but
is one of the most complete and accurate (and fast, although this is not
my major concern when wrapping it to python) genetic algorithm
implementations out there with a lot of bells and whistles for
experimentation. It also has shown a remarkably small number of bugs
over the years. It supports parallel execution via the message
passing interface <a class="reference external" href="http://mpi-forum.org/">MPI</a> in addition to a normal &quot;serial&quot; version. That's
why I wanted to use it in Python, too.</p>
<p>There currently is not much documentation for PGAPy.
You really, absolutely need to read the documentation that comes
with PGAPack – and of course you need the PGAPack library.
The PGAPack user guide is now shipped together with PGAPy. It is
installed together with some examples in share/pgapy, wherever the
Python installer decides to place this in the final installation
(usually <tt class="docutils literal">/usr/local/share</tt> on Linux).</p>
<p>The original PGAPack library can still be downloaded from the <a class="reference external" href="http://ftp.mcs.anl.gov/pub/pgapack/">PGAPack</a>
ftp site, it is written in ANSI C and therefore <em>should</em> run on most
platforms. Note that this version is not very actively maintained. I've
started a <a class="reference external" href="https://github.com/schlatterbeck/pgapack">PGAPack fork on github</a> where I've ported the library to the
latest version of the <a class="reference external" href="http://mpi-forum.org/">MPI</a> standard and have fixed some minor
inconsistencies in the documentation. I've also implemented some new
features (notably enhancements in selection schemes and a new replacement
strategy called <em>restricted tournament replacement</em>)</p>
<p>I have tested pgapy on Linux only and I'll currently not provide Windows
versions.  You also can find my <a class="reference external" href="https://github.com/schlatterbeck/pgapack">PGAPack fork on github</a> this
repository has the three upstream releases as versions in git and
contains some updates concerning support of newer <a class="reference external" href="http://mpi-forum.org/">MPI</a> versions and
documentation updates.  I've also included patches in the git repository
of the Debian maintainer of the package, Dirk Eddelbuettel.</p>
<p>To get you started, I've included some very simple examples in
<tt class="docutils literal">examples</tt>, e.g., <tt class="docutils literal"><span class="pre">one-max.py</span></tt> implements the &quot;Maxbit&quot; example
similar to one in the PGAPack documentation. The examples were inspired
by the book &quot;Genetic Algorithms in Python&quot; but are written from scratch
and don't include any code from the book. The examples illustrates
several points:</p>
<ul class="simple">
<li>Your class implementing the genetic algorithm needs to inherit from
pga.PGA (pga is the PGAPy wrapper module).</li>
<li>You need to define an evaluation function called <tt class="docutils literal">evaluate</tt> that
returns a number indicating the fitness of the gene given with the
parameters <tt class="docutils literal">p</tt> and <tt class="docutils literal">pop</tt> that can be used to fetch allele values from
the gene using the <tt class="docutils literal">get_allele</tt> method, for more details refer to the
PGAPack documentation.</li>
<li>You <em>can</em> define additional functions overriding built-in functions
of the PGAPack library, illustrated by the example of
<tt class="docutils literal">print_string</tt>.  Note that we could call the original print_string
method of our PGA superclass.  In the same way you can implement,
e.g., your own crossover method.</li>
<li>The constructor of the class needs to define the Gene type, in the
examples we use int and bool built-in datatypes.</li>
<li>The length of the gene needs to be given in the constructor.</li>
<li>We often want to maximize the numbers returned by our evaluation
function, set the parameter <tt class="docutils literal">maximize</tt> to False if you want to
minimize.</li>
<li>For non-binary genes we can define an array of init values, each entry
containing a sequence with lower and upper bound. The array has to
have the length of the gene. Note that the upper bound is <em>included</em>
in the range of possible values (unlike the python range operator but
compatible with the PGAPack definition).</li>
<li>In the constructor of the class we can add parameters of the genetic
algorithm. Not all parameters of PGAPack are wrapped yet, currently
you would need to consult the sourcecode of PGAPy to find out which
parameters are wrapped. In the example we define several print
options.</li>
<li>Finally the genetic algorithm is started with the <tt class="docutils literal">run</tt> method.</li>
</ul>
<div class="section" id="naming-conventions-in-pgapy">
<h1>Naming conventions in PGAPy</h1>
<p>When you extend PGAPy – remember not all functions of PGAPack are
wrapped yet and you may need additional functions – you should stick to
my naming conventions when making changes.
The following naming conventions were used for the wrapper:</p>
<ul class="simple">
<li>Constants of PGAPack like <tt class="docutils literal">PGA_REPORT_STRING</tt> are used as-is in
uppercase. These constants can be directly imported from the wrapper
module. Not all constants are wrapped so far, if you need more, add
them to the constdef array in pgamodule.c and <a class="reference external" href="mailto:rsc&#64;runtux.com">send</a> me a patch.</li>
<li>For methods of the pga.PGA class I've removed the <tt class="docutils literal">PGA</tt> prefix used
throughout PGAPack and converted the method to lowercase with
underscores between uppercase words in the original function name, so
<tt class="docutils literal">PGARun</tt> becomes <tt class="docutils literal">run</tt>, <tt class="docutils literal">PGACheckStoppingConditions</tt> becomes
<tt class="docutils literal">check_stopping_conditions</tt>. An exception of the lowercase-rule is
whenever a name contains &quot;GA&quot; (for &quot;genetic algorithm&quot;), So
<tt class="docutils literal">PGASetMaxGAIterValue</tt> becomes <tt class="docutils literal">max_GA_iter</tt>.</li>
<li>Where possible I've made a single class method where PGAPack needs a
separate function for each datatype, so <tt class="docutils literal">PGAGetBinaryAllele</tt>,
<tt class="docutils literal">PGAGetCharacterAllele</tt>, <tt class="docutils literal">PGAGetIntegerAllele</tt>, <tt class="docutils literal">PGAGetRealAllele</tt> all
become <tt class="docutils literal">get_allele</tt>. Same holds true for <tt class="docutils literal">set_allele</tt>.</li>
<li>Whenever a name in PGApack has a &quot;Value&quot; or &quot;Flag&quot; suffix, I've left
this out, so <tt class="docutils literal">PGAGetFitnessCmaxValue</tt> becomes <tt class="docutils literal">fitness_cmax</tt>
and <tt class="docutils literal">PGAGetMutationAndCrossoverFlag</tt> becomes
<tt class="docutils literal">mutation_and_crossover</tt>, the only exception to this rule is for the
two functions <tt class="docutils literal">PGAGetMutationRealValue</tt> and
<tt class="docutils literal">PGAGetMutationIntegerValue</tt> which become <tt class="docutils literal">mutation_value</tt> not
just <tt class="docutils literal">mutation</tt>.</li>
<li>Some fields can take multiple values (they are implemented by ORing
integer constants, in python they are specified as a list or tuple of
constants). These are converted to plural (if not already plural in
PGApack), e.g., <tt class="docutils literal">PGASetStoppingRuleType</tt> becomes <tt class="docutils literal">stopping_rule_types</tt>.</li>
<li>Internal method names in the wrapper program have a leading PGA_ – so
the class method <tt class="docutils literal">set_allele</tt> is implemented by the C-function
<tt class="docutils literal">PGA_set_allele</tt> in <tt class="docutils literal">pgamodule.c</tt>.</li>
</ul>
</div>
<div class="section" id="constructor-parameters">
<h1>Constructor Parameters</h1>
<p>PGApack has a lot of <tt class="docutils literal">PGASet</tt> and <tt class="docutils literal">PGAGet</tt> functions for setting
parameters. These are reflected in constructor parameters on the one hand
and in read-only properties of a <tt class="docutils literal">PGA</tt> object on the other hand. The
following table gives an overview of all the original PGApack names and
the names of the python wrapper. For the PGApack name I've only listed
the <tt class="docutils literal">PGASet</tt> function, in many cases there is a corresponding
<tt class="docutils literal">PGAGet</tt> function. If a corresponding read-only property exists for a
constructor parameter this is indicated in the &quot;Prop&quot; column. In some
cases properties are missing because no corresponding <tt class="docutils literal">PGAGet</tt> function
is implemented in PGApack, in other cases returning a numeric value that
has a symbolic constant in PGApy doesn't make much sense.
The properties have the same name as the constructor parameter.
There is one Property that is no constructor parameter, namely the
<tt class="docutils literal">GA_iter</tt> property that returns the current GA generation. In the type
column I'm listing the Python type. If the type is followed by a number,
more than one item of that type is specified (a sequence in Python). Some
entries contain &quot;sym&quot;, these are integer values with a symbolic constant,
the value &quot;msym&quot; indicates that several values denoted by a list of
symbolic constants can be given. A special case are the
<tt class="docutils literal">PGASetRealInitRange</tt>, <tt class="docutils literal">PGASetRealInitPercent</tt>,
<tt class="docutils literal">PGASetIntegerInitRange</tt> functions. These take two values for <em>each
allele</em> of the gene. In python this is a sequence of 2-tuples.
Note that this means that you can have different ranges of allowed values
for each allele.</p>
<p>The first two (mandatory) constructor parameters are the type of the gene
(this takes a Python type, e.g., <tt class="docutils literal">bool</tt> for a binary genome or <tt class="docutils literal">int</tt>
for an integer genome) and the length. Note that the <tt class="docutils literal">string_length</tt> is
implicitly set with the <tt class="docutils literal">length</tt> parameter. The <tt class="docutils literal">string_length</tt> is
also available as the length of the <tt class="docutils literal">PGA</tt> object using the Python
built-in <tt class="docutils literal">len</tt> function.</p>
<table border="1" class="docutils">
<colgroup>
<col width="47%" />
<col width="40%" />
<col width="8%" />
<col width="5%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">PGApack name</th>
<th class="head">Constructor parameter</th>
<th class="head">Type</th>
<th class="head">Prop</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt class="docutils literal">PGASetCrossoverProb</tt></td>
<td><tt class="docutils literal">crossover_prob</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetCrossoverType</tt></td>
<td><tt class="docutils literal">crossover_type</tt></td>
<td>sym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetFitnessCmaxValue</tt></td>
<td><tt class="docutils literal">fitness_cmax</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetFitnessType</tt></td>
<td><tt class="docutils literal">fitness_type</tt></td>
<td>sym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGAGetGAIterValue</tt></td>
<td><tt class="docutils literal">GA_iter</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetIntegerInitPermute</tt></td>
<td><tt class="docutils literal">integer_init_permute</tt></td>
<td>int2</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetIntegerInitRange</tt></td>
<td><tt class="docutils literal">init</tt></td>
<td>&nbsp;</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMaxFitnessRank</tt></td>
<td><tt class="docutils literal">max_fitness_rank</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMaxGAIterValue</tt></td>
<td><tt class="docutils literal">max_GA_iter</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMaxNoChangeValue</tt></td>
<td><tt class="docutils literal">max_no_change</tt></td>
<td>int</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMaxSimilarityValue</tt></td>
<td><tt class="docutils literal">max_similarity</tt></td>
<td>int</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationAndCrossoverFlag</tt></td>
<td><tt class="docutils literal">mutation_and_crossover</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationBoundedFlag</tt></td>
<td><tt class="docutils literal">mutation_bounded</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationIntegerValue</tt></td>
<td><tt class="docutils literal">mutation_value</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationOrCrossoverFlag</tt></td>
<td><tt class="docutils literal">mutation_or_crossover</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationProb</tt></td>
<td><tt class="docutils literal">mutation_prob</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationRealValue</tt></td>
<td><tt class="docutils literal">mutation_value</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetMutationType</tt></td>
<td><tt class="docutils literal">mutation_type</tt></td>
<td>sym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetNoDuplicatesFlag</tt></td>
<td><tt class="docutils literal">no_duplicates</tt></td>
<td>int</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetNumReplaceValue</tt></td>
<td><tt class="docutils literal">num_replace</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetPopSize</tt></td>
<td><tt class="docutils literal">pop_size</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetPopReplaceType</tt></td>
<td><tt class="docutils literal">pop_replace_type</tt></td>
<td>sym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetPrintFrequencyValue</tt></td>
<td><tt class="docutils literal">print_frequency</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetPrintOptions</tt></td>
<td><tt class="docutils literal">print_options</tt></td>
<td>msym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetPTournamentProb</tt></td>
<td><tt class="docutils literal">p_tournament_prob</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRandomizeSelect</tt></td>
<td><tt class="docutils literal">randomize_select</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRandomSeed</tt></td>
<td><tt class="docutils literal">random_seed</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRealInitRange</tt></td>
<td><tt class="docutils literal">init</tt></td>
<td>&nbsp;</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRealInitPercent</tt></td>
<td><tt class="docutils literal">init_percent</tt></td>
<td>&nbsp;</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRestartFlag</tt></td>
<td><tt class="docutils literal">restart</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRestartFrequencyValue</tt></td>
<td><tt class="docutils literal">restart_frequency</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetRTRWindowSize</tt></td>
<td><tt class="docutils literal">rtr_window_size</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetSelectType</tt></td>
<td><tt class="docutils literal">select_type</tt></td>
<td>sym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetStoppingRuleType</tt></td>
<td><tt class="docutils literal">stopping_rule_types</tt></td>
<td>msym</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">PGASetStringLength</tt></td>
<td><tt class="docutils literal">string_length</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetTournamentSize</tt></td>
<td><tt class="docutils literal">tournament_size</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetTournamentWithReplacement</tt></td>
<td><tt class="docutils literal">tournament_with_replacement</tt></td>
<td>int</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetTruncationProportion</tt></td>
<td><tt class="docutils literal">truncation_proportion</tt></td>
<td>float</td>
<td>yes</td>
</tr>
<tr><td><tt class="docutils literal">PGASetUniformCrossoverProb</tt></td>
<td><tt class="docutils literal">uniform_crossover_prob</tt></td>
<td>float</td>
<td>yes</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="pga-object-methods">
<h1>PGA Object Methods</h1>
<p>The following are the methods that can be used during the run of the
genetic search. The <tt class="docutils literal">run</tt> method is used to start the search. This can
be used, to, e.g., set an allele during hill-climbing in a custom
<tt class="docutils literal">endofgen</tt> method. Note that some methods only apply to certain gene
types, e.g. the <tt class="docutils literal">encode_int_</tt> methods can only be used on binary
alleles (they encode an integer value as a binary or gray code
representation into the gene). Other methods take or return different
types depending on the type of gene, e.g. <tt class="docutils literal">get_allele</tt> or
<tt class="docutils literal">set_allele</tt>, they call different backend functions depending on the
gene type. With the <tt class="docutils literal">set_random_seed</tt> method, the random number
generator can be re-seeded. It is usually best to seed the generator
once at (before) the beginning by specifying <tt class="docutils literal">random_seed</tt> in the
constructor. For further details consult the user guide.</p>
<table border="1" class="docutils">
<colgroup>
<col width="39%" />
<col width="24%" />
<col width="36%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Method</th>
<th class="head">Parameters</th>
<th class="head">Return</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt class="docutils literal">check_stopping_conditions</tt></td>
<td>&nbsp;</td>
<td>True if stop should occur</td>
</tr>
<tr><td><tt class="docutils literal">encode_int_as_binary</tt></td>
<td><em>p, pop,</em>
<em>frm, to, val</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">encode_int_as_gray_code</tt></td>
<td><em>p, pop,</em>
<em>frm, to, val</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">encode_real_as_binary</tt></td>
<td><em>p, pop, frm, to</em>
<em>l, u, val</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">encode_real_as_gray_code</tt></td>
<td><em>p, pop, frm, to</em>
<em>l, u, val</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">euclidian_distance</tt></td>
<td><em>p1, pop1</em>
<em>p2, pop2</em></td>
<td>float</td>
</tr>
<tr><td><tt class="docutils literal">fitness</tt></td>
<td><em>pop</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">get_allele</tt></td>
<td><em>p, pop, index</em></td>
<td>allele value</td>
</tr>
<tr><td><tt class="docutils literal">get_best_index</tt></td>
<td><em>pop</em></td>
<td>index of best string</td>
</tr>
<tr><td><tt class="docutils literal">get_evaluation</tt></td>
<td><em>p, pop</em></td>
<td>evaluation of <em>p</em> (float)</td>
</tr>
<tr><td><tt class="docutils literal">get_evaluation_up_to_date</tt></td>
<td><em>p, pop</em></td>
<td>True if up-to-date</td>
</tr>
<tr><td><tt class="docutils literal">get_fitness</tt></td>
<td><em>p, pop</em></td>
<td>fitness of <em>p</em> (float)</td>
</tr>
<tr><td><tt class="docutils literal">get_int_from_binary</tt></td>
<td><em>p, pop, frm, to</em></td>
<td>int</td>
</tr>
<tr><td><tt class="docutils literal">get_int_from_gray_code</tt></td>
<td><em>p, pop, frm, to</em></td>
<td>int</td>
</tr>
<tr><td><tt class="docutils literal">get_iteration</tt></td>
<td>&nbsp;</td>
<td>deprecated, use <tt class="docutils literal">GA_iter</tt></td>
</tr>
<tr><td><tt class="docutils literal">get_real_from_binary</tt></td>
<td><em>p, pop,</em>
<em>frm, to, l, u</em></td>
<td>float</td>
</tr>
<tr><td><tt class="docutils literal">get_real_from_gray_code</tt></td>
<td><em>p, pop,</em>
<em>frm, to, l, u</em></td>
<td>float</td>
</tr>
<tr><td><tt class="docutils literal">random01</tt></td>
<td>&nbsp;</td>
<td>float between 0 and 1</td>
</tr>
<tr><td><tt class="docutils literal">random_flip</tt></td>
<td><em>probability</em></td>
<td>0 or 1</td>
</tr>
<tr><td><tt class="docutils literal">random_gaussian</tt></td>
<td><em>mean, stddev</em></td>
<td>float</td>
</tr>
<tr><td><tt class="docutils literal">random_interval</tt></td>
<td><em>l, r</em></td>
<td>int between l, r</td>
</tr>
<tr><td><tt class="docutils literal">random_uniform</tt></td>
<td><em>l, r</em></td>
<td>float between l, r</td>
</tr>
<tr><td><tt class="docutils literal">run</tt></td>
<td>&nbsp;</td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">select_next_index</tt></td>
<td><em>pop</em></td>
<td>index selected individual</td>
</tr>
<tr><td><tt class="docutils literal">set_allele</tt></td>
<td><em>p, pop, i, value</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">set_evaluation</tt></td>
<td><em>p, pop, value</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">set_evaluation_up_to_date</tt></td>
<td><em>p, pop, status</em></td>
<td>None</td>
</tr>
<tr><td><tt class="docutils literal">set_random_seed</tt></td>
<td><em>seed</em></td>
<td>None (use constructor!)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="user-methods">
<h1>User-Methods</h1>
<p>PGApack has the concept of user functions. These allow customization of
different areas of a genetic algorihm. In Python they are implemented as
methods that can be changed in a derived class. One of the methods that
<em>must</em> be implemented in a derived class is the <tt class="docutils literal">evaluate</tt> function
(although technically it is not a user function in PGApack). It
interprets the gene and returns an evaluation value. PGApack computes a
fitness from the raw evaluation value. For some methods an up-call into
the PGA class is possible, for some methods this is not possible (and in
most cases not reasonable). Note that for the <tt class="docutils literal">stop_cond</tt> method, the
standard check for stopping conditions can be called with:</p>
<pre class="literal-block">
self.check_stopping_conditions()
</pre>
<p>The following table lists the overridable methods with their parameters
(for the function signature the first parameter <em>self</em> is omitted). Note
that in PGApack there are additional user functions that are needed for
user-defined data types which are currently not exposed in Python. In the
function signatures <em>p</em> denotes the index of the individual and <em>pop</em>
denotes the population. If more than one individual is specified (e.g.,
for crossover) these can be followed by a number. For crossover <em>c1</em> and
<em>c2</em> denote the destination individuals (children). The <em>propability</em> for
the mutation method is a floating-point value between 0 and 1. Remember
to count the number of mutations that happen, and return that value for
the mutation method!</p>
<table border="1" class="docutils">
<colgroup>
<col width="28%" />
<col width="44%" />
<col width="18%" />
<col width="10%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Method</th>
<th class="head">Call Signature</th>
<th class="head">Return Value</th>
<th class="head">Up-Call</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt class="docutils literal">check_duplicate</tt></td>
<td><em>p1, pop1, p2, pop2</em></td>
<td>True if dupe</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">stop_cond</tt></td>
<td>&nbsp;</td>
<td>True to stop</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">crossover</tt></td>
<td><em>p1, p2, p_pop, c1, c2, c_pop</em></td>
<td>None</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">endofgen</tt></td>
<td>&nbsp;</td>
<td>None</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">evaluate</tt></td>
<td><em>p, pop</em></td>
<td>float</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">gene_difference</tt></td>
<td><em>p1, pop1, p2, pop2</em></td>
<td>float</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">initstring</tt></td>
<td><em>p, pop</em></td>
<td>None</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">mutation</tt></td>
<td><em>p, pop, propability</em></td>
<td>#mutations</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">pre_eval</tt></td>
<td><em>pop</em></td>
<td>None</td>
<td>no</td>
</tr>
<tr><td><tt class="docutils literal">print_string</tt></td>
<td><em>file, p, pop</em></td>
<td>None</td>
<td>yes</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="constants">
<h1>Constants</h1>
<p>The following PGApack constants are available:</p>
<table border="1" class="docutils">
<colgroup>
<col width="34%" />
<col width="66%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Constant</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>PGA_CROSSOVER_ONEPT</td>
<td>One-point Crossover</td>
</tr>
<tr><td>PGA_CROSSOVER_TWOPT</td>
<td>Two-point Crossover</td>
</tr>
<tr><td>PGA_CROSSOVER_UNIFORM</td>
<td>Uniform Crossover</td>
</tr>
<tr><td>PGA_FITNESSMIN_CMAX</td>
<td>Map fitness by subtracting worst</td>
</tr>
<tr><td>PGA_FITNESSMIN_RECIPROCAL</td>
<td>Map fitness via reciprocal</td>
</tr>
<tr><td>PGA_FITNESS_NORMAL</td>
<td>Linear normalization of fitness</td>
</tr>
<tr><td>PGA_FITNESS_RANKING</td>
<td>Linear fitness ranking</td>
</tr>
<tr><td>PGA_FITNESS_RAW</td>
<td>Identity fitness function</td>
</tr>
<tr><td>PGA_MUTATION_CONSTANT</td>
<td>Mutation by adding/subtracting constant</td>
</tr>
<tr><td>PGA_MUTATION_GAUSSIAN</td>
<td>Mutation by selecting from Gaussian distribution</td>
</tr>
<tr><td>PGA_MUTATION_PERMUTE</td>
<td>Mutation swaps two random genes</td>
</tr>
<tr><td>PGA_MUTATION_RANGE</td>
<td>Replace gene with uniform selection from init range</td>
</tr>
<tr><td>PGA_MUTATION_UNIFORM</td>
<td>Mutation uniform from interval</td>
</tr>
<tr><td>PGA_NEWPOP</td>
<td>Symbolic constant for new population</td>
</tr>
<tr><td>PGA_OLDPOP</td>
<td>Symbolic constant for old population</td>
</tr>
<tr><td>PGA_POPREPL_BEST</td>
<td>Population replacement best strings</td>
</tr>
<tr><td>PGA_POPREPL_PAIRWISE_BEST</td>
<td>Compare same index in old and new population</td>
</tr>
<tr><td>PGA_POPREPL_RANDOM_NOREP</td>
<td>Population replacement random no replacement</td>
</tr>
<tr><td>PGA_POPREPL_RANDOM_REP</td>
<td>Population replacement random with replacement</td>
</tr>
<tr><td>PGA_POPREPL_RTR</td>
<td>Restricted Tournament Replacement</td>
</tr>
<tr><td>PGA_REPORT_AVERAGE</td>
<td>Report average evaluation</td>
</tr>
<tr><td>PGA_REPORT_HAMMING</td>
<td>Report hamming distance</td>
</tr>
<tr><td>PGA_REPORT_OFFLINE</td>
<td>Report offline</td>
</tr>
<tr><td>PGA_REPORT_ONLINE</td>
<td>Report online</td>
</tr>
<tr><td>PGA_REPORT_STRING</td>
<td>Report the string</td>
</tr>
<tr><td>PGA_REPORT_WORST</td>
<td>Report the worst evaluation</td>
</tr>
<tr><td>PGA_SELECT_LINEAR</td>
<td>Return individuals in population order</td>
</tr>
<tr><td>PGA_SELECT_PROPORTIONAL</td>
<td>Fitness-proportional selection</td>
</tr>
<tr><td>PGA_SELECT_PTOURNAMENT</td>
<td>Binary probabilistic tournament selection</td>
</tr>
<tr><td>PGA_SELECT_SUS</td>
<td>Stochastic universal selection</td>
</tr>
<tr><td>PGA_SELECT_TOURNAMENT</td>
<td>Tournament selection</td>
</tr>
<tr><td>PGA_SELECT_TRUNCATION</td>
<td>Truncation selection</td>
</tr>
<tr><td>PGA_STOP_MAXITER</td>
<td>Stop on max iterations</td>
</tr>
<tr><td>PGA_STOP_NOCHANGE</td>
<td>Stop on max number of generations no change</td>
</tr>
<tr><td>PGA_STOP_TOOSIMILAR</td>
<td>Stop when individuals too similar</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="missing-features">
<h1>Missing Features</h1>
<p>As already mentioned, not all functions and constants of PGAPack are
wrapped yet – still for many applications the given set should be
enough. If you need additional functions, you may want to wrap these and
<a class="reference external" href="mailto:rsc&#64;runtux.com">send</a> me a patch.</p>
<p>Another feature of PGAPack is currently not implemented in the wrapper,
the usage of custom datatypes. With PGAPack you can define your own
datatypes complete with their custom implementations of the genetic
algorithm functionality like crossover, mutation, etc. I don't expect
problems implementing these, though.</p>
</div>
<div class="section" id="reporting-bugs">
<h1>Reporting Bugs</h1>
<p>Please use the <a class="reference external" href="http://sourceforge.net/tracker/?group_id=152022&amp;atid=782852">Sourceforge Bug Tracker</a>  or the <a class="reference external" href="https://github.com/schlatterbeck/pgapy/issues">Github Bug Tracker</a> and</p>
<ul class="simple">
<li>give a short description of what you think is the correct behaviour</li>
<li>give a description of the observed behaviour</li>
<li>tell me exactly what you did.</li>
<li>if you can publish your source code this makes it a lot easier to
debug for me</li>
</ul>
</div>
<div class="section" id="resources">
<h1>Resources</h1>
<p>Project information and download from <a class="reference external" href="http://sourceforge.net/projects/pgapy/">Sourceforge main page</a></p>
<p>or checkout from <a class="reference external" href="http://github.com/schlatterbeck/pgapy">Github</a></p>
<p>or directly install via pypi.</p>
</div>
<div class="section" id="changes">
<h1>Changes</h1>
<p>Version 0.8: Bugfix in real mutation</p>
<ul class="simple">
<li>Fix a core-dump in the latest pgapack</li>
</ul>
<p>Version 0.7: Major changes in wrapping</p>
<ul class="simple">
<li>Now Differential Evolution is implemented, see the minfloat example
and the user guide of pgapack.</li>
</ul>
<p>Version 0.6: Major changes in wrapping</p>
<ul class="simple">
<li>Now the wrapping uses the standard Python recommendations on how to
create a custom class.</li>
<li>Update documentation</li>
<li>Rename <tt class="docutils literal">fitness_cmax</tt> (from <tt class="docutils literal">fitness_cmax_value</tt>)</li>
<li>Better error checking of parameters</li>
</ul>
<p>Version 0.5: Bug-fix release</p>
<ul class="simple">
<li>Now the <tt class="docutils literal">setup.py</tt> works, previous version had an encoding problem</li>
<li>Wrap some minor new methods</li>
<li>Bug-fix in PGAPack truncation selection</li>
</ul>
<p>Version 0.4: Bundle PGAPack</p>
<ul class="simple">
<li>The PGAPack package is now included as a git submodule. By default we
build against this library</li>
<li>License fixes: The module long shipped a <tt class="docutils literal">COPYING</tt> file that includes
the 2-clause BSD license. But the headers of <tt class="docutils literal">setup.py</tt> and <tt class="docutils literal">pgamodule.c</tt>
still included another license. This has been corrected.</li>
</ul>
<p>Version 0.3: Feature enhancements, Bug fixes</p>
<p>Port to Python3, Python2 is still supported, license change.</p>
<ul class="simple">
<li>C-Code of wrapper updated to support both, Python2 and Python3</li>
<li>Update documentation</li>
<li>Fix some memory leaks that could result when errors occurred during
some callback methods</li>
<li>License change: We now have the 2-clause BSD license (similar to the
MPICH license of PGAPack), this used to be LGPL.</li>
</ul>
<p>Version 0.2: Feature enhancements, Bug fixes</p>
<p>64 bit support, more PGAPack functions and attributes wrapped,
Readme-update: Sourceforge logo, Changes chapter.</p>
<ul class="simple">
<li>Bug-fixes for 64 bit architectures</li>
<li>More functions and attributes of PGAPack wrapped</li>
<li>Add a build-rule to <tt class="docutils literal">setup.py</tt> to allow building for standard-install
of PGAPack – this currently needs editing of <tt class="docutils literal">setup.py</tt> – should use
autodetect here but this would require that I set up a machine with
standard install of PGAPack for testing.</li>
<li>Add Sourceforge logo as required</li>
<li>Add Changes chapter for automagic releases</li>
<li>Add the <tt class="docutils literal">__module__</tt> string to class <tt class="docutils literal">PGA</tt> in module <tt class="docutils literal">pga</tt>. Now
calling:: <tt class="docutils literal">help (pga)</tt> in python works as expected, previously no
help-text was given for the included module</li>
</ul>
<p>Version 0.1: Initial freshmeat announcement</p>
<p>PGAPy is a wrapper for PGAPack, the parallel genetic algorithm library,
a powerful genetic algorithm library. PGAPy wraps this library for use
with Python. Pgapack is one of the most complete and accurate genetic
algorithm implementations out there with a lot of features for
experimentation.</p>
<ul class="simple">
<li>Initial Release</li>
</ul>
</div>
</div>
</body>
</html>
